core/src/main/scala/kafka/network/PrioritizationAwareBlockingQueue.scala

```diff
@@ -0,0 +1,139 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.network
+
+import java.util.ArrayDeque
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.{Condition, ReentrantLock}
+
+/**
+ * A blocking queue that is prioritization-aware.
+ *
+ * Prioritized elements in the queue get polled before regular elements in the queue.
+ */
+private[network] class PrioritizationAwareBlockingQueue[E](private val capacity: Int) {
+  private var count: Int = 0
```

Comment:
It looks that the count is shared by the prioritized and normal requests. When there are too many normal requests, it seems that the prioritized request may not really be prioritized to be put into the queue? So the queue is actually prioritized on the get side. This is probably OK for most of the cases. But if the queue is moving slowly and there are N putting threads, each thread may have to wait for N/2 on average. Do we want to consider prioritize on the put side as well, i.e. adding another set of conditional variables and count.

Reply:
This is an interesting point. Request prioritization would be most useful when the system is under heavy load, meaning it would be most useful when the queue is already full.

I was hoping to avoid a config change. The current implementation tries to provide a solution while still honoring the existing "queued.max.requests" config.

Reply:
Oh I think I see what you mean. We can add prioritization to the put and still honor the the existing "queued.max.requests".

Reply:
If we actually want prioritization on both the get-side and put-side, I have some code ready. It's just a matter of deciding which behavior we want.

Reply:
@onurkaraman Sorry for the late reply. I think it may worth prioritize on the put side as well. Could you update the patch with that logic?

Reply:
Just added the commit to the PR.